runtime/src/accounts_db.rs

```diff
@@ -3794,101 +3857,132 @@ impl AccountsDb {
         (num_new_roots, num_roots_flushed)
     }
 
-    // `should_flush_f` is an optional closure that determines whether a given
-    // account should be flushed. Passing `None` will by default flush all
-    // accounts
-    fn flush_slot_cache(
+    fn do_flush_slot_cache(
         &self,
         slot: Slot,
+        slot_cache: &SlotCache,
         mut should_flush_f: Option<&mut impl FnMut(&Pubkey, &AccountSharedData) -> bool>,
     ) -> FlushStats {
         let mut num_purged = 0;
         let mut total_size = 0;
         let mut num_flushed = 0;
-        let did_flush = if let Some(slot_cache) = self.accounts_cache.slot_cache(slot) {
-            let iter_items: Vec<_> = slot_cache.iter().collect();
-            let mut purged_slot_pubkeys: HashSet<(Slot, Pubkey)> = HashSet::new();
-            let mut pubkey_to_slot_set: Vec<(Pubkey, Slot)> = vec![];
-            let (accounts, hashes): (Vec<(&Pubkey, &AccountSharedData)>, Vec<Hash>) = iter_items
-                .iter()
-                .filter_map(|iter_item| {
-                    let key = iter_item.key();
-                    let account = &iter_item.value().account;
-                    let should_flush = should_flush_f
-                        .as_mut()
-                        .map(|should_flush_f| should_flush_f(key, account))
-                        .unwrap_or(true);
-                    if should_flush {
-                        let hash = iter_item.value().hash();
-                        total_size += (account.data().len() + STORE_META_OVERHEAD) as u64;
-                        num_flushed += 1;
-                        Some(((key, account), hash))
-                    } else {
-                        // If we don't flush, we have to remove the entry from the
-                        // index, since it's equivalent to purging
-                        purged_slot_pubkeys.insert((slot, *key));
-                        pubkey_to_slot_set.push((*key, slot));
-                        num_purged += 1;
-                        None
-                    }
-                })
-                .unzip();
-
-            let is_dead_slot = accounts.is_empty();
-            // Remove the account index entries from earlier roots that are outdated by later roots.
-            // Safe because queries to the index will be reading updates from later roots.
-            self.purge_slot_cache_pubkeys(
+        let iter_items: Vec<_> = slot_cache.iter().collect();
```

Comment:
I know this wasn't changed, but is this copy still needed? Could the loop a few lines down use `slot_cache` directly, instead of `iter_items`?

Reply:
@brooksprumo yeah, I think you're right, it's the more efficient thing to do.

There's a few changes you would have to make though.  The iter() returns locked entries called `RefMulti` from within the `DashMap`. If we don't create a vector of these `RefMulti'`, then for the lifetimes to work, we can't create `let (accounts, hashes): (Vec<(&Pubkey, &AccountSharedData)>, Vec<Hash>)` anymore because the `&Pubkey` references would outlive the `RefMulti` from which they were created here: https://github.com/solana-labs/solana/blob/654918ab27000b71e8af024f374e5d6e3e00cce0/runtime/src/accounts_db.rs#L3813. Instead you would have to clone out the Pubkey.

This then means when we make. this call here: https://github.com/solana-labs/solana/blob/654918ab27000b71e8af024f374e5d6e3e00cce0/runtime/src/accounts_db.rs#L3852, we have have to rework the signature of `store_accounts_frozen` https://github.com/solana-labs/solana/blob/654918ab27000b71e8af024f374e5d6e3e00cce0/runtime/src/accounts_db.rs#L5011 to support taking something like `Borrow<Pubkey>` instead of just `&Pubkey`, so that it can accept both `Pubkey` and `&Pubkey`

Might be good to tackle outside of this PR if somebody wants to play around with this stuff 😃

Reply:
Ahhh understood. Thanks for the info!